core/controllers/editor.py

```diff
@@ -960,30 +962,43 @@ def post(self, exploration_id):
                 'Invalid filename extension: it should have '
                 'one of the following extensions: %s' % allowed_formats)
 
-        temp_suffix = '.' + extension
-        with tempfile.NamedTemporaryFile(suffix=temp_suffix) as f:
-            f.write(raw)
-            try:
-                audio = mutagen.File(f.name)
-            except mutagen.MutagenError:
-                raise self.InvalidInputException('Audio not recognized '
-                                                 'as a %s file' % extension)
-            f.close()
-
+        tempbuffer = StringIO.StringIO()
+        tempbuffer.write(raw)
+        tempbuffer.seek(0)
+        try:
+            audio = mutagen.File(tempbuffer)
+        except mutagen.MutagenError:
+            # Mutagen occasionally has problems with certain files
+            # for unknown reasons.
+            raise self.InvalidInputException('Could not open uploaded'
+                                             'audio file')
+        tempbuffer.close()
+
+        if audio is None:
+            raise self.InvalidInputException('Audio not recognized '
+                                             'as a %s file' % extension)
         if audio.info.length > feconf.MAX_AUDIO_FILE_LENGTH_SEC:
             raise self.InvalidInputException(
                 'Audio must be under %s seconds in length. '
                 'Found length %s' % (feconf.MAX_AUDIO_FILE_LENGTH_SEC,
                                      audio.info.length))
+        if len(set(audio.mime).intersection(
+                set(feconf.ACCEPTED_AUDIO_EXTENSIONS[extension]))) == 0:
+            raise self.InvalidInputException(
+                'Although the filename extension indicates the file '
+                'is a %s file, it was not recognized as one. '
+                'Found mime types: %s' % (extension, audio.mime))
+
+        del audio
 
-        # Upload to GCS bucket with filepath
-        # "<bucket>/<exploration-id>/assets/audio/filename".
         bucket_name = feconf.GCS_RESOURCE_BUCKET_NAME
+
+        # Upload to GCS bucket with filepath
+        # "<bucket>/<exploration-id>/assets/audio/<filename>".
         gcs_file_url = ('/%s/%s/assets/audio/%s'
                         % (bucket_name, exploration_id, filename))
-
-        gcs_file = cloudstorage.open(gcs_file_url, 'w',
-                                     content_type=audio.mime[0])
+        gcs_file = cloudstorage.open(
+            gcs_file_url, 'w', content_type='audio/*')
```

Comment:
Okay I'll put audio.mime[0], it seems to be the most specific one, this is audio/mp3 for the test mp3 files